onnxruntime/core/session/onnxruntime_c_api.cc

```diff
@@ -2634,13 +2634,13 @@ static_assert(offsetof(OrtApi, ReleaseCANNProviderOptions) / sizeof(void*) == 22
 static_assert(std::string_view(ORT_VERSION) == "1.14.0",
               "ORT_Version change detected, please follow below steps to ensure OrtApi is updated properly");
 // 1. Update the hardcoded version string in above static_assert to silence it
-// 2. If there were any APIs added to ort_api_1_to_13 above:
+// 2. If there were any APIs added to ort_api_1_to_14 above:
 //    a. Add the 'End of version #' markers (pattern above should be obvious)
 //    b. Add a static_assert in the directly above list of version sizes to ensure nobody adds any more functions to the just shipped API version
 
 ORT_API(const OrtApi*, OrtApis::GetApi, uint32_t version) {
   if (version >= 1 && version <= ORT_API_VERSION)
-    return &ort_api_1_to_12;
+    return &ort_api_1_to_14;
```

Comment:
>return &ort_api_1_to_14; [](http://example.com/codeflow?start=4&length=24)

I am not sure we have been doing this the right way.
We are backward compatible. So the earlier version of the client program, such as C# API should function without changes. C# retrieves a complete API struct by value, but when using the newer DLL, the value is bigger.
It appears to be working, but need to check on this.

Reply:
As you say, [the complete API struct is copied by value](https://github.com/microsoft/onnxruntime/blob/main/csharp/src/Microsoft.ML.OnnxRuntime/NativeMethods.shared.cs#L277):

```c#
static NativeMethods()
{
   // ...
   // TODO: Make this save the pointer, and not copy the whole structure across
   api_ = (OrtApi)OrtGetApi(4 /*ORT_API_VERSION*/);
```

A _very brief_ look (grep for `SizeOf`) doesn't seem to indicate that we depend on the size of the API struct for anything. However, I am not familiar with the C# codebase, so I may need some help determining if this is an issue going forward.

Also, it looks like we're only querying for version 4 of the API, but we're extracting/using API functions from version 14. Should probably update that.

Reply:
I think you are right. C# is marshalling the whole struct by value though. The older code is not prepared to Marshal bigger struct. It may be okay, but a pointer would be a better option. Training API is exposed via ptr.
Not sure what happens when a newer version is requested from older ORT binary.